config/config.go

```diff
@@ -63,12 +64,62 @@ type Config struct {
 
 // Module configures a prober
 type Module struct {
-	Prober     string           `yaml:"prober,omitempty"`
-	Timeout    time.Duration    `yaml:"timeout,omitempty"`
-	TLSConfig  config.TLSConfig `yaml:"tls_config,omitempty"`
-	HTTPS      HTTPSProbe       `yaml:"https,omitempty"`
-	TCP        TCPProbe         `yaml:"tcp,omitempty"`
-	Kubernetes KubernetesProbe  `yaml:"kubernetes,omitempty"`
+	Prober     string          `yaml:"prober,omitempty"`
+	Timeout    time.Duration   `yaml:"timeout,omitempty"`
+	TLSConfig  TLSConfig       `yaml:"tls_config,omitempty"`
+	HTTPS      HTTPSProbe      `yaml:"https,omitempty"`
+	TCP        TCPProbe        `yaml:"tcp,omitempty"`
+	Kubernetes KubernetesProbe `yaml:"kubernetes,omitempty"`
+}
+
+// TLSConfig is a superset of config.TLSConfig that supports TLS renegotiation
+type TLSConfig struct {
+	CAFile             string `yaml:"ca_file,omitempty"`
+	CertFile           string `yaml:"cert_file,omitempty"`
+	KeyFile            string `yaml:"key_file,omitempty"`
+	ServerName         string `yaml:"server_name,omitempty"`
+	InsecureSkipVerify bool   `yaml:"insecure_skip_verify"`
+	// Renegotiation controls what types of TLS renegotiation are supported.
+	// Supported values: never (default), once, freely.
+	Renegotiation renegotiation `yaml:"renegotiation,omitempty"`
+}
+
+type renegotiation string
+
+func (r *renegotiation) ToRenegotiationSupport() (tls.RenegotiationSupport, error) {
```

Comment:
Shouldn’t `renegotiation` be a defined type of `tls.RenegotiationSupport` and have that logic go into `UnmarshalYAML()` instead?

Reply:
Just to be clearer, this is what I had in mind (not tested):

```suggestion
type renegotiation tls.RenegotiationSupport

func (r *renegotiation) UnmarshalYAML(unmarshal func(interface{}) error) error {
	var v string
	if err := unmarshal(&v); err != nil {
		return err
	}

	switch v {
	case "", "never":
		*r = tls.RenegotiateNever
	case "once":
		*r = tls.RenegotiateOnceAsClient
	case "freely":
		*r = tls.RenegotiateFreelyAsClient
	default:
		return fmt.Errorf("unsupported TLS renegotiation type %s", v)
	}
}
```

Reply:
That's a good idea. It means validation happens when the config file is read rather than when the scrape happens.

What do you think @britcey?

Reply:
Makes sense - I'll take a crack at it.

Reply:
Done - but I had to do a bit of type-casting in UnmarshalLYAML and then NewTLSConfig to get everything working - is that normal or am I doing something wrong?  Without it I was getting

 cannot use tls.RenegotiateNever (type tls.RenegotiationSupport) as type renegotiation in assignment

etc.

Reply:
I think that’s normal, your changes LGTM.

Reply:
LGTM2